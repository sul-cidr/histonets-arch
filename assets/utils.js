export const getCookie = (name) => {
  var cookieValue = null;
  if (document.cookie && document.cookie !== '') {
    var cookies = document.cookie.split(';');
    for (var i = 0; i < cookies.length; i++) {
      var cookie = String(cookies[i]).trim();
      if (cookie.substring(0, name.length + 1) === (name + '=')) {
        cookieValue = decodeURIComponent(cookie.substring(name.length + 1));
        break;
      }
    }
  }
  return cookieValue;
}

export const removeDuplicates = (array, key) => array
  .filter(Boolean)
  .filter((obj, pos, arr) => arr
    .map(mapObj => mapObj[key]).indexOf(obj[key]) === pos);

