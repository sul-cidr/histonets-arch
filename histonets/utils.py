import json
import re
from collections import Counter
from functools import reduce

import histonets_cv as cv

from django.conf import settings
from django.utils.text import re_camel_case


def camel_to_underscore(value):
    """
    Convert a value from camel case convention to underscore lower case convention.
    Args:
        value (str): value in camel case convention.
    Returns:
        value in underscore lowercase convention.
    """
    return re_camel_case.sub(r'_\1', value).strip('_').lower()


def underscore_to_camel(value):
    """
    Convert a value from underscore lower case convention to camel case convention.
    Args:
        value (str): value in underscore lowercase convention.
    Returns:
        value in camel case convention.
    """
    under_pat = re.compile(r'_([a-z])')
    return under_pat.sub(lambda x: x.group(1).upper(), value)


def naming_convention(dic, convert):
    """
    Convert a nested dictionary from one convention to another.
    Args:
        dic (dict): dictionary (nested or not) to be converted.
        convert (func): function that takes the string in one convention and
                        returns it in the other one.
    Returns:
        Dictionary with the new keys.
    """
    new = {}
    for key, value in dic.items():
        converted_value = value
        if isinstance(value, dict):
            converted_value = naming_convention(value, convert)
        elif isinstance(value, list):
            converted_value = []
            for each_value in value:
                if isinstance(each_value, dict):
                    converted_value.append(naming_convention(each_value,
                                                             convert))
                else:
                    converted_value.append(each_value)
        new[convert(key)] = converted_value
    return new


def to_react_props(dic):
    json_props = json.dumps(naming_convention(dic, underscore_to_camel))
    script = """<script>window.props = JSON.parse('{}')</script>"""
    return script.format(json_props)


def download_image_uri_as_array(uri):
    """Download an image from uri and turn it into a Numpy 3D RGB array."""
    return cv.utils.Image.get_images([uri])[0]


def combine_histograms(histograms):
    """Reduce by key summing up the values."""
    return dict(reduce(lambda x, y: x.update(y) or x, histograms, Counter()))
