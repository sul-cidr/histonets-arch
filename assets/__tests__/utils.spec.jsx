import { getCookie, removeDuplicates } from '../utils';

describe('utils', () => {
  it('getCookie returns a cookie value from document.cookie', () => {
    document.cookie = 'key=value;othercookie=othervalue';
    expect(getCookie('key')).toMatchSnapshot();
  });

  it('getCookie returns nothing when the cookie is not present in document.cookie', () => {
    document.cookie = 'key=value;othercookie=othervalue';
    expect(getCookie('nonexistingkey')).toMatchSnapshot();
  });

  it('getCookie returns nothing when no cookie is set in document.cookie', () => {
    document.cookie = null;
    expect(getCookie('key')).toMatchSnapshot();
  });

  it('getCookie returns nothing when cookie is an empty string', () => {
    document.cookie = '';
    expect(getCookie('key')).toMatchSnapshot();
  });

  it('removeDuplicates removes duplicates by key', () => {
    const array = [{'key': 1}, {'key': 1}, {'key': 1}, {'key': 1}];
    expect(removeDuplicates(array)).toMatchSnapshot();
  });
});
