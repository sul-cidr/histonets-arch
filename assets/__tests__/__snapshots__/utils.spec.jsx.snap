// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils getCSRFCookie returns nothing when the cookie is not present in document.cookie nor the DOM 1`] = `"token"`;

exports[`utils getCSRFCookie returns the value from the DOM when cookie is not present 1`] = `"value"`;

exports[`utils getCookie returns a cookie value from document.cookie 1`] = `"value"`;

exports[`utils getCookie returns nothing when cookie is an empty string 1`] = `"value"`;

exports[`utils getCookie returns nothing when no cookie is set in document.cookie 1`] = `"value"`;

exports[`utils getCookie returns nothing when the cookie is not present in document.cookie 1`] = `null`;

exports[`utils removeDuplicates removes duplicates by key 1`] = `
Array [
  Object {
    "key": 1,
  },
]
`;
